### DISCUSSION TITLE
Relax restrictions on attribute arguments if the attribute is used only by the compiler and analyzers

### DISCUSSION THREAD

**Discussion by dlebansais** [dotnet/csharplang#8429] [Language Ideas]
Relevant existing discussions:
[#1452](https://github.com/dotnet/csharplang/discussions/1452)
[https://github.com/dotnet/roslyn/issues/6671](https://github.com/dotnet/roslyn/issues/6671)

Currently attribute arguments are restricted to a subset of types (basically integral, strings and arrays) as explained in [this section](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#2224-attribute-parameter-types) of the documentation on attributes.

My understanding is that the CLR is limited to certain types, and so the restriction is enforced on C# attributes.

It seems to me that for attributes only used by the compiler and analyzers, if it makes sense to not put them in the resulting binary, then the restriction on types could be lifted.

Is that even possible? I do not know. Perhaps people who know more about the various steps involved in analysis and compilation can tell me more.

Is there a use case? Well, there is definitely one for me. I wrote a few analyzers, and one of them uses several attributes to tag methods and generate code. Arguments that really should be C# keywords are written as strings, then the analyzer parses the strings to do its work. I don't want to elaborate more, because my guess is that this has already been discussed somewhere (sorry, after a search all I could find are the two references above), and nobody was sufficiently interested to make a proposal for it or give consideration to the idea.

Or maybe it's just not possible, or doesn't make sense.

**Comment by jaredpar** [comment]
Think the best path forward for this is revisiting ideas like source only attributes. Essentially attributes that are only visible in source by the compiler and are never emitted to metadata. That's come up a few times ask an ask by source generator authors. If that were a feature then we could consider expanding the set of types used by them. 

That is unlikely to be a small ask though. There are likely lots of places that have an assumption about what types are allowed here. Would likely need a strong motivating example.
